src/net/tcp/stream.rs

```diff
@@ -131,23 +136,26 @@ impl TcpStream {
     /// ```
     ///
     /// [`write`]: Self::write
-    pub async fn write_all<T: IoBuf>(&self, mut buf: T) -> crate::BufResult<(), T> {
-        let mut n = 0;
-        while n < buf.bytes_init() {
-            let res = self.write(buf.slice(n..)).await;
+    pub async fn write_all<T: IoBuf>(
+        &self,
+        buf: impl Into<Slice<T>>,
+    ) -> crate::BufResult<(), Slice<T>> {
+        let mut buf = buf.into();
+        let (in_begin, in_end) = (buf.begin(), buf.end());
+        while buf.len() > 0 {
+            let res = self.write(buf).await;
             match res {
                 (Ok(0), slice) => {
                     return (
                         Err(std::io::Error::new(
                             std::io::ErrorKind::WriteZero,
                             "failed to write whole buffer",
                         )),
-                        slice.into_inner(),
+                        slice.into_inner().slice(in_begin..in_end),
                     )
                 }
-                (Ok(m), slice) => {
-                    n += m;
-                    buf = slice.into_inner();
+                (Ok(n), slice) => {
+                    buf = slice.slice(n..);
```

Comment:
I'm beginning to read slice.slice as not a dup of anything any longer. Took a few hours but now I read the first as a noun and the second as the verb acting on the noun. I'm guessing everyone else here had already been doing that.